src/test/java/io/cdap/plugin/gcp/bigquery/BigQueryRecordToJsonTest.java

```diff
@@ -225,4 +241,220 @@ public void testNullableRecord() throws IOException {
                           actual.get("base").getAsJsonObject().get("innerB").getAsJsonObject().get("int").getAsInt());
     }
   }
+
+  @Test
+  public void testNestedRecordWithJsonString() {
+    Schema nestedRecordSchema = Schema.recordOf(
+      "nestedRecord",
+            Schema.Field.of("nestedString", Schema.of(Schema.Type.STRING)),
+            Schema.Field.of("nestedJsonString", Schema.of(Schema.Type.STRING)),
+            Schema.Field.of("nestedInt", Schema.of(Schema.Type.INT))
+    );
+
+    Schema recordSchema = Schema.recordOf(
+      "record",
+            Schema.Field.of("string", Schema.of(Schema.Type.STRING)),
+            Schema.Field.of("jsonString", Schema.of(Schema.Type.STRING)),
+            Schema.Field.of("int", Schema.of(Schema.Type.INT)),
+            Schema.Field.of("nestedRecord", nestedRecordSchema)
+    );
+
+    String jsonString = "{\"string\":\"string\",\"int\":1}";
+    String nestedJsonString = "{\"nestedString\":\"nestedString\",\"nestedInt\":1}";
+
+    StructuredRecord nestedRecord = StructuredRecord.builder(nestedRecordSchema)
+      .set("nestedString", "nestedString")
+      .set("nestedJsonString", nestedJsonString)
+      .set("nestedInt", 1)
+      .build();
+
+    StructuredRecord record = StructuredRecord.builder(recordSchema)
+      .set("string", "string")
+      .set("jsonString", jsonString)
+      .set("int", 1)
+      .set("nestedRecord", nestedRecord)
+      .build();
+
+    JsonObject jsonObject = gson.fromJson(jsonString, JsonObject.class);
+    JsonObject nestedJsonObject = gson.fromJson(nestedJsonString, JsonObject.class);
+
+    Set<String> jsonStringFieldsPaths = ImmutableSet.of("jsonString", "nestedRecord.nestedJsonString");
+    try {
+      JsonTreeWriter writer = new JsonTreeWriter();
+      writer.beginObject();
+      for (Schema.Field recordField : Objects.requireNonNull(record.getSchema().getFields())) {
+        if (recordSchema.getField(recordField.getName()) != null) {
+          BigQueryRecordToJson.write(writer, recordField.getName(), record.get(recordField.getName()),
+                                     recordField.getSchema(), jsonStringFieldsPaths);
+        }
+      }
+      writer.endObject();
+
+      JsonObject actual = writer.get().getAsJsonObject();
+
+      Assert.assertEquals(jsonObject, actual.get("jsonString").getAsJsonObject());
+      Assert.assertEquals(nestedJsonObject, actual.get("nestedRecord").getAsJsonObject().get("nestedJsonString")
+        .getAsJsonObject());
+    } catch (IOException e) {
+      Assert.fail("Failed to convert record to json");
+    }
+
```

Comment:
Any particular reason of handling the exception using try-catch instead of declaring the exception?

Reply:
Updated ✅